File: rental_platform/rental_platform/otp/phoneno_edit.py
import frappe
import random
import string
from datetime import datetime
from frappe.utils import get_datetime, now_datetime, add_to_date

@frappe.whitelist()
def otp_for_new_phoneno(mobile_no=None, action=None):
    try:
        if action == "send_otp":
            if not mobile_no or len(mobile_no) != 10 or not mobile_no.isdigit():
                return {"success": False, "message": "Please enter a valid 10-digit mobile number."}
            if frappe.db.exists("Customer", {"mobile_no": mobile_no}):
                return {"success": False, "message": "Customer already exists with this mobile number."}

            otp_code = ''.join(random.choices(string.digits, k=4))  
            
            otp_verification = frappe.new_doc("OTP Verification")
            otp_verification.phone = mobile_no
            otp_verification.otp = otp_code
            otp_verification.custom_sms_type = "Onboarding"
            otp_verification.time = datetime.now()
            otp_verification.save(ignore_permissions=True)
            
            return {"success": True, "message": "OTP sent successfully.", "otp_code": otp_code}  
    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "OTP Sending Error")   
        return {"success": False, "message": str(e)}

@frappe.whitelist()
def otp_verify_new_phoneno(mobile_no=None, otp_value=None):   
    try:
        if not otp_value:
            return {"success": False, "message": "OTP cannot be empty. Please enter the OTP."}
            
        otp_record = frappe.db.get_value("OTP Verification", {"phone": mobile_no, "otp": otp_value}, ["name", "creation"])
        if not otp_record:
            return {"success": False, "message": "Invalid OTP. Please enter the correct OTP."}
        
        current_time = get_datetime(now_datetime())
        otp_creation_time = get_datetime(otp_record[1])
        otp_expiry_time = add_to_date(otp_creation_time, seconds=60)

        if current_time > otp_expiry_time:
            return {"success": False, "message": "OTP has expired. Please request a new OTP."}

        return {"success": True, "message": "OTP verified successfully."}
    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "OTP Verification Error")
        return {"success": False, "message": str(e)}


@frappe.whitelist()
def update_customer_contact(contact_name, new_mobile_no):
    try:
        contact = frappe.get_doc("Contact", contact_name)
        customer_link = frappe.db.get_value(
            "Dynamic Link",
            {"parenttype": "Contact", "link_doctype": "Customer", "parent": contact.name},
            "link_name"
        )

        if not customer_link:
            return {"success": False, "message": "No linked Customer found for this Contact."}

        if contact.is_primary_contact:
            customer = frappe.get_doc('Customer', customer_link)
            customer.mobile_no = new_mobile_no
            customer.save(ignore_permissions=True)

            # Update phone_nos child table in Contact doctype
            updated = False
            for phone_entry in contact.phone_nos:
                if phone_entry.is_primary_mobile_no:
                    phone_entry.phone = new_mobile_no
                    updated = True
                    break

            # If no primary phone entry was found, add a new one
            if not updated:
                contact.append("phone_nos", {
                    "phone": new_mobile_no,
                    "is_primary_mobile_no": 1
                })

            contact.save(ignore_permissions=True)
            frappe.db.commit()
            return {"success": True, "message": "Customer primary contact updated successfully."}

        else:
            customer = frappe.get_doc("Customer", customer_link)
            customer.custom_alternate_number = new_mobile_no
            customer.save(ignore_permissions=True)

            # Update phone_nos child table in Contact doctype
            # Update phone_nos child table in Contact doctype
                        # Update phone_nos child table in Contact doctype
            updated = False
            for phone_entry in contact.phone_nos:
                if not phone_entry.is_primary_mobile_no:  # Find the existing alternate number
                    phone_entry.phone = new_mobile_no
                    updated = True
                    break

            # If no alternate phone entry was found, return an error (instead of adding a new row)
            if not updated:
                return {"success": False, "message": "No alternate contact found to update."}

            contact.save(ignore_permissions=True)
            frappe.db.commit()
            return {"success": True, "message": "Customer alternate contact updated successfully."}

    except Exception as e:
        frappe.log_error(f"Error in update_customer_primary_contact: {str(e)}")
        return {"success": False, "message": f"An error occurred: {str(e)}"}

File: rental_platform/public/js/phoneno_edit.js
frappe.ui.form.on('Contact', {
    setup: function(frm) {
        frm._previous_phone_nos = frm.doc.phone_nos ? frm.doc.phone_nos.map(row => ({ ...row })) : [];
    },

    validate: function(frm) {
        let changed_phone = get_changed_phone_number(frm);
        if (changed_phone) {
            frappe.validated = false; // Prevent form submission
            send_otp_and_show_dialog(frm, changed_phone.phone);
        }
    }
});

function get_changed_phone_number(frm) {
    if (!frm.doc.phone_nos || !frm._previous_phone_nos) {
        return null;
    }

    for (let i = 0; i < frm.doc.phone_nos.length; i++) {
        let current_row = frm.doc.phone_nos[i];
        let previous_row = frm._previous_phone_nos[i];

        if (!previous_row || current_row.phone !== previous_row.phone) {
            return { phone: current_row.phone };
        }
    }
    return null;
}

function send_otp_and_show_dialog(frm, new_mobile_no) {
    frappe.call({
        method: 'rental_platform.rental_platform.otp.phoneno_edit.otp_for_new_phoneno',
        args: { mobile_no: new_mobile_no, action: 'send_otp' },
        callback: function(response) {
            if (response.message.success) {
                frappe.show_alert(__('OTP sent successfully to ') + new_mobile_no);
                show_otp_dialog(frm, new_mobile_no);
            } else {
                frappe.msgprint(__('Failed to send OTP: ') + response.message.message);
            }
        }
    });
}

function show_otp_dialog(frm, new_mobile_no) {
    let dialog = new frappe.ui.Dialog({
        title: __('OTP Verification'),
        fields: [
            { label: __('OTP'), fieldname: 'otp', fieldtype: 'Data', reqd: 1 },
            { label: __('Verify OTP'), fieldname: 'verify_otp', fieldtype: 'Button', click: function() {
                let otp_value = dialog.get_value('otp');
                if (!otp_value) {
                    frappe.msgprint(__('Please enter the OTP.'));
                    return;
                }

                frappe.call({
                    method: 'rental_platform.rental_platform.otp.phoneno_edit.otp_verify_new_phoneno',
                    args: { mobile_no: new_mobile_no, otp_value: otp_value },
                    callback: function(response) {
                        if (response.message.success) {
                            frappe.show_alert(__('OTP Verified Successfully!'));
                            dialog.hide();
                            update_customer_contact(frm, new_mobile_no);
                        } else {
                            frappe.msgprint(__('OTP Verification Failed: ') + response.message.message);
                        }
                    }
                });
            }}
        ]
    });

    dialog.show();
}

function update_customer_contact(frm, new_mobile_no) {
    frappe.call({
        method: 'rental_platform.rental_platform.otp.phoneno_edit.update_customer_contact',
        args: { contact_name: frm.doc.name, new_mobile_no: new_mobile_no },
        callback: function(response) {
            if (response.message && response.message.success) {
                frappe.msgprint(__('Customer contact updated successfully.'));
                frm.reload_doc();
            } else {
                frappe.msgprint(__('Failed to update customer contact: ') + response.message.message);
            }
        }
    });
}
